File: packages/pvm/src/host-functions/accumulate/transfer.ts
import { logger } from '@pbnjam/core'
import type { DeferredTransfer, HostFunctionResult } from '@pbnjam/types'
import { ACCUMULATE_FUNCTIONS, RESULT_CODES } from '../../config'
import {
  type AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
} from './base'

/**
 * TRANSFER accumulation host function (Ω_T)
 *
 * Transfers tokens between service accounts
 *
 * Gray Paper Specification:
 * - Function ID: 20 (transfer)
 * - Gas Cost: 10 + l (gasLimit) on success, 10 on error
 * - Parameters: registers[7-10] = dest, amount, l, o
 *   - dest: destination service account ID
 *   - amount: transfer amount
 *   - l: gas limit for the transfer
 *   - o: memo offset in memory
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read memo from memory (128 bytes)
 * 2. Check if destination service exists
 * 3. Check if gas limit is sufficient for destination
 * 4. Check if sender has sufficient balance
 * 5. Create deferred transfer entry
 * 6. Deduct amount from sender's balance
 */
export class TransferHostFunction extends BaseAccumulateHostFunction {
  readonly functionId = ACCUMULATE_FUNCTIONS.TRANSFER
  readonly name = 'transfer'

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const { registers, ram, implications } = context
    // Extract parameters from registers
    // Gray Paper pvm_invocations.tex line 818: [dest, amount, l, o] = registers[7:4]
    // registers[7] = dest (destination service ID)
    // registers[8] = amount (transfer amount)
    // registers[9] = l (gas limit for transfer)
    // registers[10] = o (memo offset in memory)
    const [destinationServiceId, amount, gasLimit, memoOffset] =
      registers.slice(7, 11)

    const logServiceId = implications[0].id

    // Read memo from memory (128 bytes - Gray Paper Cmemosize)
    // Gray Paper pvm_invocations.tex lines 820-832:
    // t = error when Nrange(o, Cmemosize) not readable
    // c = panic when t = error
    // registers'_7 = registers_7 (unchanged) when c = panic
    const C_MEMO_SIZE = 128n
    const [memoData, faultAddress] = ram.readOctets(memoOffset, C_MEMO_SIZE)
    if (faultAddress || !memoData) {
      // Gray Paper line 832: c = panic when t = error
      // Gray Paper line 839: registers'_7 = registers_7 (unchanged) when c = panic
      // DO NOT modify registers[7] - it must remain unchanged on panic
      logger.info(
        `[host-calls] [${logServiceId}] TRANSFER(${destinationServiceId}, ${amount}, ${gasLimit}) <- PANIC`,
      )
      return {
        resultCode: RESULT_CODES.PANIC,
      }
    }

    // Gray Paper: Gas cost is 10 + t, where t = gasLimit on success, t = 0 on error
    // We don't check OOG here - that's handled by the context mutator after we return
    // The gas cost depends on success/failure, so we just validate the base cost of 10
    // The additionalGasCost returned will be applied after this function returns

    // Get the current implications context
    const [imX] = implications

    // Get current service account
    const currentService = imX.state.accounts.get(imX.id)
    if (!currentService) {
      this.setAccumulateError(registers, 'HUH')
      logger.info(
        `[host-calls] [${logServiceId}] TRANSFER(${destinationServiceId}, ${amount}, ${gasLimit}) <- HUH`,
      )
      return {
        resultCode: null, // continue execution
      }
    }

    // Check if destination service exists
    const destService = imX.state.accounts.get(destinationServiceId)
    if (!destService) {
      this.setAccumulateError(registers, 'WHO')
      logger.info(
        `[host-calls] [${logServiceId}] TRANSFER(${destinationServiceId}, ${amount}, ${gasLimit}) <- WHO`,
      )
      return {
        resultCode: null, // continue execution
      }
    }

    // Check if gas limit is sufficient for destination
    // Gray Paper: l < destService.sa_minmemogas
    if (gasLimit < destService.minmemogas) {
      this.setAccumulateError(registers, 'LOW')
      logger.info(
        `[host-calls] [${logServiceId}] TRANSFER(${destinationServiceId}, ${amount}, ${gasLimit}) <- LOW`,
      )
      return {
        resultCode: null, // continue execution
      }
    }

    // Check if sender has sufficient balance after transfer
    // Gray Paper line 830: b = (imX_self)_sa_balance - amount
    // Gray Paper line 835: CASH when b < (imX_self)_sa_minbalance
    // Gray Paper accounts.tex: sa_minbalance = max(0, Cbasedeposit + Citemdeposit * items + Cbytedeposit * octets - gratis)
    const balanceAfterTransfer = currentService.balance - amount

    // Calculate minbalance according to Gray Paper accounts.tex
    const C_BASEDEPOSIT = 100n
    const C_ITEMDEPOSIT = 10n
    const C_BYTEDEPOSIT = 1n
    const baseDeposit = C_BASEDEPOSIT
    const itemDeposit = C_ITEMDEPOSIT * currentService.items
    const byteDeposit = C_BYTEDEPOSIT * currentService.octets
    const totalDeposit = baseDeposit + itemDeposit + byteDeposit
    const minbalance =
      totalDeposit > currentService.gratis
        ? totalDeposit - currentService.gratis
        : 0n

    if (balanceAfterTransfer < minbalance) {
      this.setAccumulateError(registers, 'CASH')
      logger.info(
        `[host-calls] [${logServiceId}] TRANSFER(${destinationServiceId}, ${amount}, ${gasLimit}) <- CASH`,
      )
      return {
        resultCode: null, // continue execution
      }
    }

    // Create deferred transfer entry
    // Gray Paper: t = {source: imX.id, dest, amount, memo, gas: l}
    const deferredTransfer = {
      source: imX.id,
      dest: destinationServiceId,
      amount,
      memo: memoData,
      gasLimit,
    } satisfies DeferredTransfer

    // Add transfer to xfers list
    imX.xfers.push(deferredTransfer)

    // Deduct amount from sender's balance
    // CRITICAL: Ensure we're modifying the account that's actually in the state map
    // Get the account directly from the state map to ensure we have the correct reference
    const accountInState = imX.state.accounts.get(imX.id)
    if (!accountInState) {
      this.setAccumulateError(registers, 'HUH')
      logger.info(
        `[host-calls] [${logServiceId}] TRANSFER(${destinationServiceId}, ${amount}, ${gasLimit}) <- HUH`,
      )
      return {
        resultCode: null, // continue execution
      }
    }

    // Verify we have the same reference (should always be true, but check for safety)
    if (accountInState !== currentService) {
      logger.warn(
        '[TransferHostFunction] Account reference mismatch, using account from state map',
        {
          serviceId: imX.id.toString(),
        },
      )
    }

    accountInState.balance = balanceAfterTransfer

    // Set success result
    this.setAccumulateSuccess(registers)

    // Log in the requested format: [host-calls] [serviceId] TRANSFER(dest, amount, gasLimit) <- OK
    logger.info(
      `[host-calls] [${logServiceId}] TRANSFER(${destinationServiceId}, ${amount}, ${gasLimit}) <- OK`,
    )

    // Gray Paper: On success, gas cost is 10 + l (where l = gasLimit)
    // The base 10 gas is deducted in the context mutator, so we return
    // the gasLimit as additionalGasCost to be deducted there
    return {
      resultCode: null, // continue execution
      additionalGasCost: gasLimit,
    }
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/transfer.rs
//! TRANSFER accumulation host function (Ω_T). Gray Paper: function ID 20.
//! r7..r10 = dest, amount, gas limit (l), memo offset (o). 1:1 with AS transfer.ts.
//! Read memo (128 bytes); current account → HUH if missing; dest → WHO if missing;
//! gasLimit < dest.minmemogas → LOW; balance checks → CASH; else deduct balance, r7=OK, gas 10+l.

use crate::codec::DeferredTransfer;
use crate::config::{C_BASE_DEPOSIT, C_BYTE_DEPOSIT, C_ITEM_DEPOSIT, C_MEMO_SIZE, FUNC_TRANSFER};
use crate::host_functions::accumulate::base::{self, codes};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

/// On success total gas = 10 + gasLimit; base 10 deducted in state_wrapper, we deduct gasLimit here.

pub struct TransferHostFunction;

impl TransferHostFunction {
    /// Gray Paper accounts.tex: minbalance = max(0, Cbasedeposit + Citemdeposit*items + Cbytedeposit*octets - gratis)
    fn minbalance(account: &crate::codec::CompleteServiceAccount) -> u64 {
        let total_deposit = C_BASE_DEPOSIT
            .saturating_add(C_ITEM_DEPOSIT * account.items as u64)
            .saturating_add(C_BYTE_DEPOSIT * account.octets);
        total_deposit.saturating_sub(account.gratis)
    }
}

impl HostFunction for TransferHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_TRANSFER
    }
    fn name(&self) -> &'static str {
        "transfer"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let dest = context.registers[7];
        let amount = context.registers[8];
        let gas_limit = context.registers[9];
        let memo_offset = context.registers[10];

        let _log_service_id = context.service_id.unwrap_or(0);

        // Read memo from memory (128 bytes). Gray Paper: panic when range not readable; r7 unchanged.
        let read_result = context.ram.read_octets(memo_offset as u32, C_MEMO_SIZE);
        if read_result.fault_address != 0 || read_result.data.is_none() {
            crate::host_log!(
                "[hostfn] transfer PANIC: memo read fault (offset={}, fault_address={})",
                memo_offset, read_result.fault_address
            );
            crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- PANIC", _log_service_id, dest, amount, gas_limit);
            return HostFunctionResult::panic();
        }
        let _memo_data = read_result.data.unwrap();
        if _memo_data.len() != C_MEMO_SIZE as usize {
            crate::host_log!(
                "[hostfn] transfer PANIC: memo length mismatch (got {}, expected {})",
                _memo_data.len(),
                C_MEMO_SIZE as usize
            );
            return HostFunctionResult::panic();
        }

        let service_id = match context.service_id {
            Some(id) => id as u32,
            None => {
                base::set_accumulate_error(context.registers, codes::HUH);
                crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- HUH", _log_service_id, dest, amount, gas_limit);
                return HostFunctionResult::continue_execution();
            }
        };

        let accounts = match context.accounts.as_deref_mut() {
            Some(a) => a,
            None => {
                base::set_accumulate_error(context.registers, codes::WHO);
                crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- WHO", _log_service_id, dest, amount, gas_limit);
                return HostFunctionResult::continue_execution();
            }
        };
        // 1. Check if source service exists (HUH check)
        // We need to check existence before resolving dest (WHO check) to match AS order.
        let source_exists = context.service_account.is_some()
            || accounts.contains_key(&(service_id as u64));

        if !source_exists {
            base::set_accumulate_error(context.registers, codes::HUH);
            crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- HUH", _log_service_id, dest, amount, gas_limit);
            return HostFunctionResult::continue_execution();
        }

        // 2. Check if destination service exists (WHO check)
        let dest_minmemogas = match accounts.get(&dest) {
            Some(s) => s.minmemogas,
            None => {
                base::set_accumulate_error(context.registers, codes::WHO);
                crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- WHO", _log_service_id, dest, amount, gas_limit);
                return HostFunctionResult::continue_execution();
            }
        };

        // 3. Acquire mutable reference to source account
        // Should always succeed because we checked source_exists above
        let current_account = match &mut context.service_account {
            Some(acc) => acc,
            None => match accounts.get_mut(&(service_id as u64)) {
                Some(acc) => acc,
                None => {
                    // This should be unreachable given the check above
                    base::set_accumulate_error(context.registers, codes::HUH);
                    crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- HUH", _log_service_id, dest, amount, gas_limit);
                    return HostFunctionResult::continue_execution();
                }
            },
        };

        // Gray Paper: l < destService.minmemogas → LOW
        if gas_limit < dest_minmemogas {
            base::set_accumulate_error(context.registers, codes::LOW);
            crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- LOW", _log_service_id, dest, amount, gas_limit);
            return HostFunctionResult::continue_execution();
        }

        // balance < amount → CASH
        if current_account.balance < amount {
            base::set_accumulate_error(context.registers, codes::CASH);
            crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- CASH", _log_service_id, dest, amount, gas_limit);
            return HostFunctionResult::continue_execution();
        }
        let balance_after_transfer = current_account.balance - amount;

        // balanceAfterTransfer < minbalance → CASH
        let minbalance = Self::minbalance(current_account);
        if balance_after_transfer < minbalance {
            base::set_accumulate_error(context.registers, codes::CASH);
            crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- CASH", _log_service_id, dest, amount, gas_limit);
            return HostFunctionResult::continue_execution();
        }

        // Success: deduct amount from sender and record deferred transfer for implications.xfers.
        current_account.balance = balance_after_transfer;
        if let Some(xfers) = &mut context.xfers {
            xfers.push(DeferredTransfer {
                source: service_id,
                dest: dest as u32,
                amount,
                memo: _memo_data.to_vec(),
                gas_limit,
            });
        }
        base::set_accumulate_success(context.registers, codes::OK);
        crate::host_log!("[host-calls] [{}] TRANSFER({}, {}, {}) <- OK", _log_service_id, dest, amount, gas_limit);

        // Gray Paper: on success additional gas = l (base 10 already deducted in state_wrapper)
        let additional_gas = gas_limit as u32;
        *context.gas_remaining = context.gas_remaining.saturating_sub(additional_gas);

        HostFunctionResult::continue_execution()
    }
}
